### PR TITLE
Ruby: Context sensitive instance method resolution

### PR DESCRIPTION
**Pull request by hvitved**
This PR implements call-sensitive instance method resolution, using [the data-flow library's built-in support](https://github.com/github/codeql/blob/main/docs/ql-libraries/dataflow/dataflow.md#virtual-dispatch-with-call-context).

For example, in
```rb
class A
  def method1 x
    sink x
  end

  def method2 x
    method1 x
  end
end

class B < A
  def method1 x
    non_sink x
  end
end

B.new.method2(taint)
```
we will no longer report flow from `taint` to `sink` (but still to `non_sink`).

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/controlflow/CfgNodes.qll [hunk 1]
```diff
@@ -444,6 +444,9 @@ module ExprNodes {
 
     /** Gets the body of this `when`-clause. */
     final ExprCfgNode getBody() { e.hasCfgChild(e.getBody(), this, result) }
+
+    /** Gets the `i`th pattern this `when`-clause. */
```
Comment (intrigus-lgtm) [inline ruby/ql/lib/codeql/ruby/controlflow/CfgNodes.qll:448]:
```suggestion
    /** Gets the `i`th pattern of this `when`-clause. */
```

File: ruby/ql/lib/codeql/ruby/ast/Module.qll [hunk 2]
```diff
@@ -14,6 +14,9 @@ class Module extends TModule {
   /** Gets the super class of this module, if any. */
   Module getSuperClass() { result = getSuperClass(this) }
 
+  /** Gets a sub class of this module, if any. */
+  Module getASubClass() { this = getSuperClass(result) }
```
Comment (aibaars) [inline ruby/ql/lib/codeql/ruby/ast/Module.qll:18]: This predicate returns direct sub classes only, doesn't it? Let document that and perhaps also make it more clear from its name.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll [hunk 3]
```diff
@@ -438,90 +448,103 @@ private DataFlow::LocalSourceNode trackModuleAccess(Module m) {
...
-          not m instanceof SingletonMethod and
-          if m.getEnclosingModule() instanceof Toplevel then exact = true else exact = false
-        )
-        or
-        selfInToplevel(self.getVariable(), tp) and
-        exact = true
-      )
+  or
+  // `in C => c then c.foo`
+  asModulePattern(n, tp) and
```
Comment (aibaars) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll:535]: Perhaps also add `C === x` tests. If I'm not mistaken, both types of `case` expressions are implemented using that operator.

Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll:535]: Ah I see you just moved these predicate. Perhaps we should post-pone `===` for a future PR. We might want to desugar `case` statements into `if` at some point too.

**Review by aibaars — COMMENTED** [review]
Looks good to me.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll (hunk 3 above)
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll:535]: Yes, we can do that in another PR. I guess is could be added to `hasAdjacentTypeCheckedReads`.

**Review by aibaars — APPROVED** [review]

**Pull request merged**